share/scripts/sdcdb.tcl

```diff
@@ -0,0 +1,795 @@
+            set addr2file([expr {"0x$address"}]) [list $filename[ASM] $linenum]
+            debug_out "mapping asm source ${arrayname}\($linenum\): [set ${arrayname}($linenum)]"
+            incr a_count
+            continue
+        }
+        set match [regexp -inline $func_bn_pat $line]
+        if {[llength $match] == 8} {
+            lassign $match {} context {} {} funcname {} {} address
+            # Put function begin record in array with dynamic name
+            switch -- [string index $context 0] {
+                G {
+                    variable g_func2addr
+                    if {[complete g_func2addr $funcname [list [expr {"0x$address"}]]]} {
+                        debug_out "1:mapping function g_func2addr\($funcname\): [set g_func2addr($funcname)]"
+                        incr gf_count
+                    }
+                }
+                F {
+                    set arrayname [string range $context 1 [string length $context]]_func2addr
+                    variable $arrayname
+                    if {[complete $arrayname $funcname [list [expr {"0x$address"}]]]} {
+                        debug_out "1:mapping function ${arrayname}\($funcname\): [set ${arrayname}($funcname)]"
+                        incr sf_count
+                    }
+                }
+                L {}
+            }
+            continue
+        }
+        set match [regexp -inline $func_ed_pat $line]
+        if {[llength $match] == 8} {
+            lassign $match {} context {} {} funcname {} {} address
+            # Put function end record in array with dynamic name
+            switch -- [string index $context 0] {
+                G {
+                    variable g_func2addr
+                    if {[complete g_func2addr $funcname [list [expr {"0x$address"}]]]} {
+                        debug_out "mapping function g_func2addr\($funcname\): [set g_func2addr($funcname)]"
+                    }
+                }
+                F {
+                    set arrayname [string range $context 1 [string length $context]]_func2addr
+                    variable $arrayname
+                    if {[complete $arrayname $funcname [list [expr {"0x$address"}]]]} {
+                        debug_out "mapping function ${arrayname}\($funcname\): [set ${arrayname}($funcname)]"
+                    }
+                }
+                L {
+                }
+            }
+            continue
+        }
+        debug_out "Ignored '$line'"
+    }
+    close $fh
+    variable c_files
+    variable a_files
+    puts "[array size c_files] C files added"
+    puts "[array size a_files] assembly files added"
+    puts "[array size c_files_count] C files references added"
+    puts "[array size a_files_count] assembly files references added"
+    puts "$c_count C source lines found"
+    puts "$a_count assembly source lines found"
+    puts "$gf_count global function(s) registered"
+    puts "$sf_count static function(s) registered"
```

Comment:
Question: should these remain `puts` or should these be returned as a (multi-line) string?